.github/workflows/release.yml

```diff
@@ -0,0 +1,74 @@
+name: Release RubyGems
+
+on:
+  push:
+    branches:
+      - release/*
+
+jobs:
+
+  pre-release-check:
+    name: Perform checks before releasing
+    runs-on: ubuntu-latest
+
+    steps:
+      - uses: actions/checkout@v2
+
+      - name: Check commit has been pushed on origin/main
+        run: |
+          git fetch --quiet origin main
+          git merge-base --is-ancestor HEAD origin/main
+
+  tests:
```

Comment:
I would prefer not to have to run tests on release, if the commit has already been tested before.

It's nice if the workflow can run as quickly as possible, and it should not be necessary to test a commit a second time if it's already been tested. I *think* we can ensure this with some checkboxes in the GitHub branch protection rules, but I'll do some experimentation today with that to verify my assumption.

Reply:
OK, I've done some testing with this. The way the "Require status checks to pass before merging" protection rule seems to work, you have to be _merging commits into the branch_ for this rule to kick in.

The way we've been doing it, by simply pushing a new `refs/heads` pointing to the commit, doesn't seem to count as a merge, so it's happy to let you do that. You can make the rule kick in by starting the branch with an earlier commit, and then merging in the newer one:

```
$ git push origin main^:refs/heads/release/v1.2.9
Total 0 (delta 0), reused 0 (delta 0), pack-reused 0
remote:
remote: Create a pull request for 'release/v1.2.9' on GitHub by visiting:
remote:      https://github.com/cucumber/release-tests/pull/new/release/v1.2.9
remote:
To github.com:cucumber/release-tests.git
 * [new branch]      main^ -> release/v1.2.9
$ git checkout release/v1.2.9
Branch 'release/v1.2.9' set up to track remote branch 'release/v1.2.9' from 'origin'.
Switched to a new branch 'release/v1.2.9
$ git merge --ff-only main
Updating 9bd6964..e193f30
Fast-forward
 .github/workflows/test.yml | 4 +---
 1 file changed, 1 insertion(+), 3 deletions(-)
$ git push
Total 0 (delta 0), reused 0 (delta 0), pack-reused 0
remote: error: GH006: Protected branch update failed for refs/heads/release/v1.2.9.
remote: error: Required status check "Run tests" is failing.
To github.com:cucumber/release-tests.git
 ! [remote rejected] release/v1.2.9 -> release/v1.2.9 (protected branch hook declined)
error: failed to push some refs to 'github.com:cucumber/release-tests.git'
```

This seems like a flaw at GitHub's end, but whatever: it's possible to push a commit to a release branch that's untested (or fails tests), so we should run the tests before we release.

Reply:
> 
> I would prefer not to have to run tests on release, if the commit has already been tested before.
> 
> It's nice if the release workflow can run as quickly as possible, and it should not be necessary to test a commit a second time if it's already been tested. I _think_ we can ensure this with some checkboxes in the GitHub branch protection rules, but I'll do some experimentation today with that to verify my assumption.

At some point we have to wait for the tests being executed on the CI. If it is nor part of the release itself, it is just before.

In the release process, we do a release commit, then push it on `main` and on `release`.

Without executing the tests as part of the release process, we do the release commit, push it on `main`, wait for the tests passing, push to `release`, wait for the release to be published.

We wait two dinstinct times. This is disruptive.

When executing the tests as part of the release process, we can push to `main` and to `release` at the same time safely. Then we wait a single time for the gem to be published with all the tests being executed just before.

Actually I prefer to wait a single time, maybe a little bit longer, than waiting, two distinct times.

This is a small repo, with a short test-suite, as cucumber-expressions is. When we'll reach the point where we have to adapt the release process to a much bigger repo, with longer tests, we may have to find some ways to make it faster. But I think we are not there yet.

Reply:
Yeah, that's a good point.

I think this is definitely good enough for now.